src/connection/host_lookup.c

```diff
@@ -0,0 +1,411 @@
+
+      do
+	{
+	  if (*token == '\0' || *token == '#')
+	    {
+	      break;
+	    }
+	  if (!storage_flag)
+	    {
+	      strcpy (hostent_List[host_count].addr, token);
+	      storage_flag = true;
+	    }
+	  else
+	    {
+	      strcpy (hostent_List[host_count].hostname, token);
+	    }
+	}
+      while (token = strtok_r (NULL, delim, &save_ptr_strtok));
+//naming rule check, detailed rule, error set
+//if hostent_List[host_count].hostname is empty, set error
+      host_count++;
+    }
+  return host_count;
+}
+
+/*
+ * gethostbyname_uhost () - Do same job with gethostbyname (), using by the 'user' defined 'hosts.conf' file.
+ * 
+ * return   : the hostent pointer.
+ * hostname (in) : the hostname.
+ *
+ */
+hostent *
+gethostbyname_uhost (char *hostname)
+{
+  hostent *hp;
+
+  if (host_conf_Use < 0)
+    {
+      host_conf_Use = prm_get_bool_value (PRM_ID_USE_USER_HOSTS);
+    }
+
+  if (host_conf_Use == 0)
+    {
+      return gethostbyname (hostname);
+    }
+
+  if (!host_conf_element_Count)
+    {
+      host_conf_element_Count = host_conf_load ();
+      if (host_conf_element_Count <= 0)
+	{
+	  return NULL;
+	}
+    }
+
+  hp = host_lookup_internal (hostname, NULL, HOSTNAME_LOOKUP);
+
+  if (hp == NULL)
+    {
+      return NULL;
+    }
+
+  return hp;
+}
+
+int
+gethostbyname_r_uhost (const char *hostname, struct hostent *out_hp)
+{
+  struct hostent *hp_memcpy = NULL;
+  int ret;
+
+  if (host_conf_Use < 0)
+    {
+      host_conf_Use = prm_get_bool_value (PRM_ID_USE_USER_HOSTS);
+    }
+
+  if (host_conf_Use == 0)
+    {
+//err also modified by callee
+#ifdef HAVE_GETHOSTBYNAME_R
+#if defined (HAVE_GETHOSTBYNAME_R_GLIBC)
+      struct hostent *hp, hent;
+      int herr;
+      char buf[1024];
+
+      ret = gethostbyname_r (hostname, &hent, buf, sizeof (buf), &hp, &herr);
+#elif defined (HAVE_GETHOSTBYNAME_R_SOLARIS)
+      struct hostent hent;
+      int herr;
+      char buf[1024];
+
+      ret = gethostbyname_r (hostname, &hent, buf, sizeof (buf), &herr);
+#elif defined (HAVE_GETHOSTBYNAME_R_HOSTENT_DATA)
+      struct hostent hent;
+      struct hostent_data ht_data;
+
+      ret = gethostbyname_r (hostname, &hent, &ht_data);
+#else
+#error "HAVE_GETHOSTBYNAME_R"
+#endif
+#else /* HAVE_GETHOSTBYNAME_R */
+      struct hostent *hent;
+      int r;
+
+      r = pthread_mutex_lock (&gethostbyname_lock);
+      hent = gethostbyname (hostname);
+      if (hent == NULL)
+	pthread_mutex_unlock (&gethostbyname_lock);
+#endif /* !HAVE_GETHOSTBYNAME_R */
+      memcpy (out_hp, &hent, sizeof (hent));
+      out_hp = &hent;
+      return ret;
+    }
+
+  if (!host_conf_element_Count)
+    {
+      host_conf_element_Count = host_conf_load ();
+      if (host_conf_element_Count <= 0)
+	{
+	  return NULL;
+	}
+    }
+
+  hp_memcpy = host_lookup_internal (hostname, NULL, HOSTNAME_LOOKUP);
+  memcpy ((void *) out_hp, (void *) hp_memcpy, sizeof (struct hostent));
+  //gethostbyname_r has different return values by OS
+
+  return 0;
```

Comment:
always returns 0 is not good, please handle ERROR cases (refer error code in manual gethostbyname_r manual entry)

Reply:
I agree with your opinion. It will be fixed when gethostbyname_r_uhost is fixed